test_conformance/api/test_api_consistency.cpp

```diff
@@ -0,0 +1,501 @@
+// limitations under the License.
+//
+#include "testBase.h"
+#include "harness/testHarness.h"
+
+const char* test_kernel = R"CLC(
+__kernel void test(__global int* dst) {
+    dst[0] = 0;
+}
+)CLC";
+
+int test_consistency_svm(cl_device_id deviceID, cl_context context,
+                         cl_command_queue queue, int num_elements)
+{
+    // clGetDeviceInfo, passing CL_DEVICE_SVM_CAPABILITIES:
+    // May return 0, indicating that device does not support Shared Virtual
+    // Memory.
+    int error;
+
+    const size_t allocSize = 16;
+    clMemWrapper mem;
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+
+    cl_device_svm_capabilities svmCaps = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_SVM_CAPABILITIES,
+                            sizeof(svmCaps), &svmCaps, NULL);
+    test_error(error, "Unable to query CL_DEVICE_SVM_CAPABILITIES");
+
+    if (svmCaps == 0)
+    {
+        mem =
+            clCreateBuffer(context, CL_MEM_READ_WRITE, allocSize, NULL, &error);
+        test_error(error, "Unable to create test buffer");
+
+        error = create_single_kernel_helper(context, &program, &kernel, 1,
+                                            &test_kernel, "test");
+        test_error(error, "Unable to create test kernel");
+
+        // clGetMemObjectInfo, passing CL_MEM_USES_SVM_POINTER
+        // Returns CL_FALSE if no devices in the context associated with
+        // memobj support Shared Virtual Memory.
+        cl_bool usesSVMPointer;
+        error =
+            clGetMemObjectInfo(mem, CL_MEM_USES_SVM_POINTER,
+                               sizeof(usesSVMPointer), &usesSVMPointer, NULL);
+        test_error(error, "Unable to query CL_MEM_USES_SVM_POINTER");
+
+        // Check that the SVM APIs can be called.
+
+        // Returns NULL if no devices in context support Shared Virtual Memory.
+        void* ptr0 = clSVMAlloc(context, CL_MEM_READ_WRITE, allocSize, 0);
+        void* ptr1 = clSVMAlloc(context, CL_MEM_READ_WRITE, allocSize, 0);
+        if (ptr0 != NULL || ptr1 != NULL)
+        {
+            log_error("CL_DEVICE_SVM_CAPABILITIES returned 0 but clSVMAlloc "
+                      "returned a non-NULL value\n");
+            return TEST_FAIL;
+        }
+
+        // clEnqueueSVMFree, clEnqueueSVMMemcpy, clEnqueueSVMMemFill,
+        // clEnqueueSVMMap, clEnqueueSVMUnmap, clEnqueueSVMMigrateMem Returns
+        // CL_INVALID_OPERATION if the device associated with command_queue does
+        // not support Shared Virtual Memory.
+
+        cl_uint pattern = 0xAAAAAAAA;
+        error = clEnqueueSVMMemFill(queue, ptr0, &pattern, sizeof(pattern),
+                                    allocSize, 0, NULL, NULL);
+        if (error != CL_INVALID_OPERATION)
+        {
+            log_error(
+                "CL_DEVICE_SVM_CAPABILITIES returned 0 but "
+                "clEnqueueSVMMemFill did not return CL_INVALID_OPERATION\n");
+            return TEST_FAIL;
+        }
```

Comment:
Can make these checks a little more concise by using `test_failure_error`, e.g.:

```
test_failure_error(error, CL_INVALID_OPERATION,
                   "CL_DEVICE_SVM_CAPABILITIES returned 0 but "
                   "clEnqueueSVMMemFill did not return CL_INVALID_OPERATION\n");
```

Reply:
Cool, I was looking for something like this but didn't find it.  I've switched all of these checks to use `test_failure_error`.

It would be nice to have a similar function when checking for a specific value (but not a specific error code).  I looked for this one but I didn't find it.  Does it exist?  I'm thinking of something like `test_value` that could be used to simplify e.g.:

```c
if (u != 0)
{
    log_error(
        "CL_DEVICE_PIPE_SUPPORT returned CL_FALSE, but "
        "CL_DEVICE_PIPE_MAX_PACKET_SIZE returned a non-zero value\n");
    return TEST_FAIL;
}
```

to:

```c
test_value(u, 0,
           "CL_DEVICE_PIPE_SUPPORT returned CL_FALSE, but "
           "CL_DEVICE_PIPE_MAX_PACKET_SIZE returned a non-zero value");
```

Reply:
I agree, I think it would be useful to add a function that can test arbitrary expressions like this, and I don't believe it exists. I'd suggest something that takes a boolean expression as the first argument to make it usable in more cases, e.g.:
```
// Or test_assert() ?
test_value(ptr0 != NULL || ptr1 != NULL,
           "CL_DEVICE_SVM_CAPABILITIES returned 0 but clSVMAlloc "
           "returned a non-NULL value");
```

It should be a pretty straightforward addition to `errorHelpers.h`. I don't think this change would have to hold up this PR though.

Reply:
Sounds good, I'll take a look at this for "Part 2".

(I like the idea of switching to a boolean expression.  Maybe `test_condition` instead of `test_value` if we do this though?  I'd be fine with `test_assert` also.)